types/handler.go

```diff
@@ -40,14 +40,18 @@ func ChainAnteDecorators(chain ...AnteDecorator) AnteHandler {
 		return nil
 	}
 
-	// handle non-terminated decorators chain
-	if (chain[len(chain)-1] != Terminator{}) {
-		chain = append(chain, Terminator{})
+	handlerChain := make([]AnteHandler, len(chain)+1)
+	// Install the terminal AnteHandler.
+	handlerChain[len(chain)] = func(ctx Context, tx Tx, simulate bool) (Context, error) {
+		return ctx, nil
 	}
-
-	return func(ctx Context, tx Tx, simulate bool) (Context, error) {
-		return chain[0].AnteHandle(ctx, tx, simulate, ChainAnteDecorators(chain[1:]...))
+	for i := 0; i < len(chain); i++ {
+		ii := i
```

Comment:
why are you setting an internal variable?
```suggestion
```

Reply:
The loop variable `i` would be captured by reference by the closure which is why we need use the local `ii`. See the `Closures` section of https://eli.thegreenplace.net/2019/go-internals-capturing-loop-variables-in-closures/ for more details.

Reply:
Ahh yes, I missed the func closure. Indeed that is correct.